gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/replication/ConfigBasedDataset.java

```diff
@@ -177,12 +231,30 @@ public String datasetURN() {
         watermarkMetadataCopied = true;
       }
 
-      // skip copy same file
-      if (copyToFileMap.containsKey(newPath) && copyToFileMap.get(newPath).getLen() == originFileStatus.getLen()
-          && copyToFileMap.get(newPath).getModificationTime() > originFileStatus.getModificationTime()) {
-        log.debug("Copy from timestamp older than copy to timestamp, skipped copy {} for dataset with metadata {}",
-            originFileStatus.getPath(), this.rc.getMetaData());
-      } else {
+
+      boolean shouldCopy = true;
+      if (copyToFileMap.containsKey(newPath)) {
+        if (this.srcDataFileVersionStrategy.get() instanceof ModTimeDataFileVersionStrategy) {
+          // default version strategy should also compare data file length
+          if (copyToFileMap.get(newPath).getLen() == originFileStatus.getLen()
+              && copyToFileMap.get(newPath).getModificationTime() > originFileStatus.getModificationTime()) {
+            log.debug("Copy from timestamp older than copy to timestamp, skipped copy {} for dataset with metadata {}",
+                originFileStatus.getPath(), this.rc.getMetaData());
+            shouldCopy = false;
+          }
+        } else {
+          // other version strategy should just compare the version number
+          Comparable srcVer = this.srcDataFileVersionStrategy.get().getVersion(originFileStatus.getPath());
+          Comparable dstVer = this.dstDataFileVersionStrategy.get().getVersion(copyToFileMap.get(newPath).getPath());
+          if (srcVer.compareTo(dstVer) == 0) {
```

Comment:
You want to copy only if `srcVersion > dstVersion`, right?

Reply:
I think so, let me change it.